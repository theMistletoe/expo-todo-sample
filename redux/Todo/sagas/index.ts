/* eslint-disable no-constant-condition */
import { put, takeEvery, call } from 'redux-saga/effects'

export function fetchBuzzWordApi() {
  return fetch('https://corporatebs-generator.sameerkumar.website/')
    .then(response => response.json())
}

type BuzzWord = {
  "phrase": string;
}

export function* fetchBuzzWord() {
  try {
    const buzzWord: BuzzWord = yield call(fetchBuzzWordApi);
    yield put({
      type: 'ADD_TASK',
      title: buzzWord.phrase,
    });
  } catch (e) {
    // TODO 未実装
    yield put({type: "USER_FETCH_FAILED", message: e.message});
  }
}

export default function* rootSaga() {
  yield takeEvery('ADD_TASK_FROM_API', fetchBuzzWord);
}
