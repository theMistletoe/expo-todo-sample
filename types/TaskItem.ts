export type TaskItem = {
    title: string;
    description: string;
};