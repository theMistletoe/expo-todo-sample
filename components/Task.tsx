import { useNavigation } from '@react-navigation/native';
import React from 'react';
import { StyleSheet, Text } from 'react-native';
import { TaskItem } from '../types/TaskItem';

type Props = {
    task: TaskItem;
    index: number;
}

const Task = (props: Props) => {
    const navigation = useNavigation(); // (2)
    // TODO: deleteボタンの実装
    const handlePressOnTaskTitle = () => {
      navigation.navigate(
        'Detail',
        {
          id: props.task.id,
        }
      )
    };

    return (
      <Text
        style={styles.item}
        key={props.index}
        onPress={handlePressOnTaskTitle}
      >
        {props.task.title}
      </Text>
    );
};

const styles = StyleSheet.create({
    item: {
        padding: 10,
        fontSize: 18,
        height: 44,
    },
});

export default Task;