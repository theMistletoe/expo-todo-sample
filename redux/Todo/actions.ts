import { TaskItem } from '../../types/TaskItem';

export const ADD_TASK = "ADD_TASK";
export const DELETE_TASK = "DELETE_TASK";
export const UPDATE_TASK = "UPDATE_TASK";

export const updateTask = (task: TaskItem) => ({
  type: 'UPDATE_TASK',
  task
});


export const deleteTask = (id: string) => ({
  type: 'DELETE_TASK',
  id
});

export const addTask = (title:string) => ({
  type: 'ADD_TASK',
  title: title,
});
