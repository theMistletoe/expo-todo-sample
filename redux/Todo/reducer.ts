import { TaskItem } from '../../types/TaskItem';

const INITIAL_STATE: TaskItem[] = [
  {id: "xxx", title: 'Task1', description: "description1"},
  {id: "yyy",title: 'Task2', description: "description2"},
  {id: "zzz",title: 'Task3', description: "description3"},
]
const reducer = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case 'ADD_TASK':
      const newTask: TaskItem = {
        id: Math.random().toString(32).substring(2),
        title: action.title,
        description: ""
      };
      return [...state, newTask];
    case 'DELETE_TASK':
      const newTasks: TaskItem[] = state.filter((todo: TaskItem) => todo.id !== action.id);
      return newTasks
    case 'UPDATE_TASK':
      const updatedTasks: TaskItem[] = state.slice();
      updatedTasks.find((task: TaskItem) => task.id === action.task.id)!.title = action.task.title;
      updatedTasks.find((task: TaskItem) => task.id === action.task.id)!.description = action.task.description;
      return updatedTasks
    default:
      return state;
  }
}

export default reducer;