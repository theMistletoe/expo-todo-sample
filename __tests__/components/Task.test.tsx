import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import Task from '../../components/Task';
import { TaskItem } from '../../types/TaskItem';
import { deleteTask } from '../../redux/Todo/actions';

const mockDispatch = jest.fn();
jest.mock("react-redux", () => ({
  useSelector: jest.fn(),
  useDispatch: () => mockDispatch
}));

const mockNavigate = jest.fn()
jest.mock('@react-navigation/native', () => {
  return {
    ...jest.requireActual('@react-navigation/native') as any,
    useNavigation: () => ({
      navigate: mockNavigate
    }),
  }
})

const sampleTask: TaskItem = {
  id: "xxx",
  title: "titie",
  description: "desc"
};

describe("test Task", () => {
  it('renders correctly with defaults', () => {
    const comp = renderer.create(<Task task={sampleTask} index={0} />)
    expect(comp).toMatchSnapshot()
  })

  it('navigate to desc page by pressing Title', () => {
    const { getByText } = render(<Task task={sampleTask} index={0} />);
    fireEvent.press(getByText("titie"));
    expect(mockNavigate).toHaveBeenCalledWith(
      'Detail',
      {
        id: "xxx",
      }
    );
  })

  it('delete task by pressing Delete Button and navigate to Home', () => {
    const { getByText } = render(<Task task={sampleTask} index={0} />);
    fireEvent.press(getByText("Delete"));
    expect(mockDispatch).toHaveBeenCalledWith(deleteTask("xxx"));
    expect(mockNavigate).toHaveBeenCalledWith("Home");
  })
})