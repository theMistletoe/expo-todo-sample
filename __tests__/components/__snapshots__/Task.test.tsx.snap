// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Task renders correctly with defaults 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "flexDirection": "row",
    }
  }
>
  <Text
    onPress={[Function]}
    style={
      Object {
        "fontSize": 18,
        "height": 44,
        "padding": 10,
      }
    }
  >
    titie
  </Text>
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Delete
      </Text>
    </View>
  </View>
</View>
`;
